hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/AutoRecordKeyGenerationUtils.scala

```diff
@@ -0,0 +1,103 @@
+      HoodieTableConfig.AUTO_GENERATE_RECORD_KEYS.defaultValue()).toBoolean
+
+    if (autoGenerateRecordKeys) {
+      // check for supported operations.
+      if (!supportedOperations.contains(operation.name())) {
+        throw new HoodieException(operation.name() + " is not supported with Auto generation of record keys. "
+          + "Supported operations are : " + supportedOperations)
+      }
+      // de-dup is not supported with auto generation of record keys
+      if (parameters.getOrElse(HoodieWriteConfig.COMBINE_BEFORE_INSERT.key(),
+        HoodieWriteConfig.COMBINE_BEFORE_INSERT.defaultValue()).toBoolean) {
+        throw new HoodieException("Enabling " + HoodieWriteConfig.COMBINE_BEFORE_INSERT.key() + " is not supported with auto generation of record keys ");
+      }
+      // drop dupes is not supported
+      if (hoodieConfig.getBoolean(INSERT_DROP_DUPS)) {
+        throw new HoodieException("Enabling " + INSERT_DROP_DUPS.key() + " is not supported with auto generation of record keys ");
+      }
+      // virtual keys are not supported with auto generation of record keys.
+      if (!parameters.getOrElse(HoodieTableConfig.POPULATE_META_FIELDS.key(), HoodieTableConfig.POPULATE_META_FIELDS.defaultValue().toString).toBoolean) {
+        throw new HoodieException("Disabling " + HoodieTableConfig.POPULATE_META_FIELDS.key() + " is not supported with auto generation of record keys");
+      }
+    }
+  }
+
+  /**
+   * Auto Generate record keys when auto generation config is enabled.
+   * <ol>
+   *   <li>Generated keys will be unique not only w/in provided [[org.apache.spark.sql.DataFrame]], but
+   *   globally unique w/in the target table</li>
+   *   <li>Generated keys have minimal overhead (to compute, persist and read)</li>
+   * </ol>
+   *
+   * Keys adhere to the following format:
+   *
+   * [instantTime]_[PartitionId]_[RowId]
+   *
+   * where
+   * instantTime refers to the commit time of the batch being ingested.
+   * PartitionId refers to spark's partition Id.
+   * RowId refers to the row index within the spark partition.
+   *
+   * @param autoGenerateKeys true if auto generation of record keys is enabled. false otherwise.
+   * @param genRecsItr Iterator of GenericRecords.
+   * @param instantTime commit time of the batch.
+   * @return Iterator of Pair of GenericRecord and Optionally generated record key.
+   */
+  def mayBeAutoGenerateRecordKeys(autoGenerateKeys : Boolean, genRecsItr: Iterator[GenericRecord], instantTime: String): Iterator[(GenericRecord, Option[String])] = {
+    var rowId = 0
+    val sparkPartitionId = TaskContext.getPartitionId()
+
+    // we will override record keys if auto generation if keys is enabled.
+    genRecsItr.map(avroRecord =>
+      if (autoGenerateKeys) {
+        val recordKey : String = HoodieRecord.generateSequenceId(instantTime, sparkPartitionId, rowId)
+        rowId += 1
```

Comment:
Also, does it need to be atomic counter like we do for commit seq number in create/append handles?

Reply:
since these are executed from within a single executor. I didn't find any need. might confuse developers later if we change to Atomic. so, would prefer to keep it this way.

Reply:
We should check why it is Atomic in some of the write handles, eg. HoodieAppendHandle. If it's not needed then we can do away with this overhead.